Light filters

mattigruener opened the pull request:
Hey John,

another one - I thought we'd get the ball rolling on the light filter stuff :) Not sure if you'll be too chuffed with some of the decisions I've made here, especially the tagging via metadata.

Not tagging these as ai:surface has fixed the crashes btw. Are you OK with pushing the postfixing into the base class like I did here? Thought that might be useful for a bunch of other things as well...

Next on my list are visualisers and making the backend translate the shader assignments properly.

Let me know what you want fixed :)

Cheerio for now!

Matti.

johnhaddon commented:
Thanks Matti,

> another one - I thought we'd get the ball rolling on the light filter stuff :) Not sure if you'll be too chuffed with some of the decisions I've made here, especially the tagging via metadata.

Light filters are a bit weird aren't they? Most of them seem to be identified by having no output type, but then light_decay goes and ruins it by having a float output. I also found an Arnold [tutorial](https://support.solidangle.com/display/AFMUG/Creating+a+Light+Filter+Shader) where they write a light filter shader and register it with an RGB output type. I really don't think this makes much sense though - light filters write into the `sg->Liu` light intensity global, so I don't see how they could also be used elsewhere in a network. Also, that example doesn't seem to actually write to its RGB output. Might be worth asking SolidAngle what the deal is, or checking to see if things have improved in 5?

Anyway, in the absence of other information, I don't see any decent way of dealing with this other than with metadata of some sort - it looks like MtoA uses "maya.classification" for this purpose.

> Not tagging these as ai:surface has fixed the crashes btw. Are you OK with pushing the postfixing into the base class like I did here? Thought that might be useful for a bunch of other things as well...

I'm not sure yet about pushing this into the base class, but I _am_ pretty sure that using the node name to generate the suffix isn't the way to go. Admittedly we do use it already for generating more-easily-debuggable networks, but that's in the role of blind data only, whereas this pushes things to the point where it actually changes data that makes it to the renderer. The difference between node names and plug values is already something that tends to confuse folks, so I don't think we want to blur the lines here any more - node names are merely for organisation, whereas plug values are for computation. As one example of where this can go wrong, imagine copy-pasting a node network alongside the original - the copy now generates an entirely different set of attribute names (because the nodes were renamed), which is pretty unexpected.

I guess some other options are :

1. Just expose the `typePlug()`  to the user so they can change the name if they want to. This is nice because it's simple, but it seems wrong to give the user the chance to mess up the needed "ai:lightFilter:" prefix.
2. Keep the suffix plug on the base class and expose it directly to the user so they can use it instead of the node name. This seems nicer from a user interaction perspective, but it does seem a bit odd to have it on the base class when using it would mess up surface shaders and the like. I guess we could hide it from the UI for surface shaders pretty easily though.
3. Add the suffix plug on the ArnoldShader only when a light filter is loaded, and overload `ArnoldShader::attributes()` to use it. Then do the same for RenderMan light filters and anything else that crops up.
4. Resolve duplicate attribute names at the point of assignment, so we rename light filters there so they don't clash with the existing ones.
5. Punt for now, and restrict things to one of each light filter type.
6. Back up and say that the whole "one attribute per filter" idea was rubbish.

I think I'm leaning towards 2, but I've included a few more radical options in case they give you any bright ideas. Out of interest, what is the bunch of other things you think the base class functionality could be used for? Seems like a good use case could tip things towards 2?

> Let me know what you want fixed :)

Mostly I think we need to figure out the question of how we generate the attribute suffixes, but I noticed the indentation was off in a number of places too. Somewhere I have some git settings that highlight indentation issues before you commit them - I'll try and fish them out if you're interested...

Cheers...
John

mattigruener commented:
Hey John :)

> Light filters are a bit weird aren't they? Most of them seem to be identified by having no output type, but then light_decay goes and ruins it by having a float output.

They sure are! I also thought we might get away with just looking at the output... To be honest - I like tagging them explicitly. I don't know what other shaders might come along without having an output type and tagging the light filters seems like a good way to defend against that (improbable?) case... 

Regarding using the node name for the shader handle part in ai:lightFilter:handle, - I guess what I was thinking is that the node name is a simple way to get a somewhat unique identifier so that copy/pasting actually creates new filters instead of overriding the previous ones. I get what you're saying though and we should find a better way. 

I am also leaning towards (2). I guess another solution would be

7. (2) + Make _every_ shader have a handle like that and have the rest of the code ignore it, if it doesn't use it.

Not sure how much would have to be changed in existing code. It seems to me that a new renderer:type:name convention could make sense in general and the fact that some code chokes on the 'name' part then is that code's fault maybe? I'm happy to go with (2) the way you outlined it, of course.

> Out of interest, what is the bunch of other things you think the base class functionality could be used for?

I guess [currently we can't write light filters in OSL](https://support.solidangle.com/display/A5ARP/OSL+Shaders):

> Arnold supports most OSL shader globals, like P, u, v, N, Ng and time. Their meaning is the same as in the C++ shading API.
> Ps (for light filters), surfacearea(), dtime and dPdtime shader globals are not supported currently.

But if light filters could be written in OSL, that would be an argument to have the postfixing in a common base class? I guess what I was thinking about initially was addressing shaders via a string handle like we'd do for co-shaders in RIB, but in Gaffer that doesn't make a whole lot of sense. Not sure if we can think of anything else where shader assignment is done through string handles in set expressions... 

If you're ok with going with (2), I'll go ahead and expose the plug and add some UI magic.

Thanks for having a look!

mattigruener commented:
I fixed the indentation problems and added more commits that implement a working prototype for assigning the filters and have them translated to Arnold. Some bits are added in anticipation of the location based filters and are therefore untested. This will need some unittests and I'll get to that tomorrow, I think, but I wanted to get it into this PR in case you have a chance to look at this before our meeting tomorrow - no problem if you don't :)

Cheerio!

johnhaddon commented:
>> Light filters are a bit weird aren't they? Most of them seem to be identified by having no output type, but then light_decay goes and ruins it by having a float output.

> To be honest - I like tagging them explicitly. I don't know what other shaders might come along without having an output type and tagging the light filters seems like a good way to defend against that (improbable?) case... 

Agreed. I wonder though if it might make more sense to have a "shaderType" metadata set to "lightFilter" than the boolean "isLightFilter" metadata? We can reuse the "shaderType" form if some other weird type comes along in the future, whereas the boolean approach doesn't really provide us anywhere to go.

> Regarding using the node name for the shader handle part in ai:lightFilter:handle, - I guess what I was thinking is that the node name is a simple way to get a somewhat unique identifier so that copy/pasting actually creates new filters instead of overriding the previous ones.

Yeah, I can see the logic in that, but I think the basic principle of names and plugs being different things with different purposes has to take precedence here.

> I get what you're saying though and we should find a better way. I am also leaning towards (2).

Cool. Unless we're going to shake things up big time I think that's probably our best bet.

> 7. (2) + Make _every_ shader have a handle like that and have the rest of the code ignore it, if it doesn't use it.
>
>Not sure how much would have to be changed in existing code. It seems to me that a new >renderer:type:name convention could make sense in general and the fact that some code chokes on the 'name' part then is that code's fault maybe?

I don't really understand how this could make sense in general. For most shader types that can be assigned (light, surface, displacement), you can only assign one at a time, so allowing multiple attributes with different suffixes doesn't make sense - it creates ambiguity. These light filters are the first example of something where we want multiple assignments of the same type to coexist, which is what makes the suffix meaningful...

>> Out of interest, what is the bunch of other things you think the base class functionality could be used for?

> If light filters could be written in OSL, that would be an argument to have the postfixing in a common base class?

Yeah, and we'll want the same thing when we come to do PRMan light filters. It does seem a bit weird still, but I think I can live with this being on the base class.

> If you're ok with going with (2), I'll go ahead and expose the plug and add some UI magic.

Yeah, let's give it a bash. We're just adding an `attributeSuffixPlug()` now right? We won't need the second "useAttributeSuffix" boolean any more? I'm saying "suffix" rather than "postfix" because a bit of grepping shows that that's the term we've been using so far in Gaffer...

> Some bits are added in anticipation of the location based filters and are therefore untested.

Let's keep the location based filters out of this PR, and concentrate on doing a polished implementation of the assignment based filters first. Like wot we agreed on the email :)

Cheers...
John

mattigruener commented:
There's still a little problem with the UI that I will address today. I'm just commenting to let you know I'm aware of it :)

mattigruener commented:
Not sure if you like my UI modifications to `ArnoldShaderUI`. That one is a little special in how it sets everything up and just adding my stuff in the end seemed a little odd. Do you want me to do that differently?

We also get this warning when deserializing gobos:
`WARNING : GafferArnold::ParameterHandler::setupPlug : Unsupported parameter "rotate" of type "ARRAY" on node "gobo"`

It's a little annoying because it pops up a dialog when opening a file. We should probably try to get rid of it before encouraging people to use the gobo node? I'm not sure if there is a plan in place for the array type? :)

johnhaddon commented:
> Not sure if you like my UI modifications to ArnoldShaderUI. That one is a little special in how it sets everything up and just adding my stuff in the end seemed a little odd. Do you want me to do that differently?

Seems good to me.

> We also get this warning when deserializing gobos:
WARNING : GafferArnold::ParameterHandler::setupPlug : Unsupported parameter "rotate" of type "ARRAY" on node "gobo"

I'm presuming that the rotate parameter is an array so as to support motion blur. Since we don't support motion blurred shader parameters, and this is effectively a shader parameter, the best we could do is load it as a float instead. Perhaps we should allow a `gaffer.plugType STRING FloatPlug` metadata value to request that, and then ensure that the Arnold backend can cope with creating arrays from scalars? Or, if we're looking for a quick fix we could just register `gaffer.plugType STRING ""` as we have with other parameters. I suspect the latter would just be a short term thing though, since folks could reasonably expect to have the rotate parameter work.

mattigruener commented:
I'm happy to put in some more work on this one. Just so I understand what you propose regarding the `FloatPlug` solution - if we replace the plug with a `FloatPlug`, would you expect the backend to convert it to `[ f ]`? I'm not sure what Arnold expects as valid values for that plug - just the angle or a full transform. I checked mtoa and exported an .ass file from Maya which resulted in that plug just not being exported. The UI also is not very helpful and doesn't let users set the rotation at all. Maybe - at least in mtoa - they expect people to rotate the light instead?

johnhaddon commented:
Did a bit of experimenting with this in an [ass file](https://github.com/GafferHQ/gaffer/files/1131422/test.txt). Here's what I found :

- It rotates the slide map around the axis of the light, as you might expect
- It seemed to accept a single float without having to convert to an array manually (this could just be the ass parser being helpful, but it's worth trying)
- It also accepted an array of 2 floats to motion blur the rotation. But it assumes that the shutter range is always 0-1, which is not the case in files generated from gaffer

So it seems like it might be possible to just create a FloatPlug for it and have it trickle through, and it also seems that there's no real point trying to support motion blur on it.

mattigruener commented:
Having it trickle through seems to work.

Are you ok with what I've done to your Color3f/4f override mechanism? I don't think I broke it - tests are still passing.

johnhaddon commented on src/GafferArnold/ArnoldShader.cpp:
```diff
@@ -128,12 +128,37 @@ void ArnoldShader::loadShader( const std::string &shaderName, bool keepExistingV
+	{
+		bool isLightFilter = false;
+		const CompoundData *metadata = ArnoldShader::metadata();
+		if( metadata )
```
Is it actually possible to get NULL metadata? As far as I can tell it isn't, in which case let's not defend against it.

johnhaddon commented on src/GafferArnold/ArnoldShader.cpp:
```diff
@@ -128,12 +128,37 @@ void ArnoldShader::loadShader( const std::string &shaderName, bool keepExistingV
+			const StringData *shaderTypeData = static_cast<const StringData*>( metadata->member<IECore::CompoundData>( "shader" )->member<IECore::Data>( "shaderType" ) );
+			if( shaderTypeData )
+			{
+				isLightFilter = shaderTypeData->readable() == "lightFilter";
```
When I suggested the "shaderType" metadata my intention was that it be a more generic mechanism than the "isLightFilter" mechanism we started with, but this is undoing that. Can we not just say `type = "ai:" + shaderTypeData->readable()`?

johnhaddon commented on include/GafferScene/Shader.h:
```diff
@@ -142,6 +142,9 @@ class Shader : public Gaffer::DependencyNode
 		virtual void attributesHash( const Gaffer::Plug *output, IECore::MurmurHash &h) const;
 		virtual IECore::ConstCompoundObjectPtr attributes( const Gaffer::Plug *output ) const;
 
+		Gaffer::StringPlug *attributeSuffixPlug();
```
I think we're intending the user to be able to modify the values of this plug directly aren't we? If that's the case then I think it should be public, up there just before the parameters plug.

johnhaddon commented on python/GafferArnoldTest/ArnoldShaderTest.py:
```diff
@@ -363,6 +363,24 @@ def testAssignmentAttributeName( self ) :
+		p = GafferScene.Plane()
+
+		s = GafferArnold.ArnoldShader( "light_blocker" )
+		s.loadShader( "light_blocker" )  # has isLightFilter tag in metadata
```
Outdated comment.

johnhaddon commented on python/GafferArnoldTest/ArnoldShaderTest.py:
```diff
@@ -363,6 +363,24 @@ def testAssignmentAttributeName( self ) :
+		a["in"].setInput( p["out"] )
+		a["shader"].setInput( s["out"] )
+
+		self.assertEqual( s["attributeSuffix"].getValue(), "" )
```
Question - should the attribute suffix be set to the name of the shader when we load it? That way I could assign a gobo and a blocker without them conflicting with each other, and would only need to use the suffix to allow two gobos to coexist. It would also mean that all assignment would match the pattern "ai:lightFilter:*" which I think is what we originally intended?

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -47,6 +47,7 @@
 
 #include "boost/make_shared.hpp"
 #include "boost/format.hpp"
+#include <boost/algorithm/string.hpp>
```
Please match the "" use from the surrounding code. If there's a case to be made for using <>, then make it and we can change everything, but let's keep consistency.

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -533,13 +535,20 @@ class ArnoldAttributes : public IECoreScenePreview::Renderer::AttributesInterfac
+				if( boost::starts_with( it->first.string(), g_lightFilterPrefix.string() ) )
 				{
-					m_user[it->first] = data;
+					if( it->second )
```
Any good reason to test against NULL here? As far as I'm aware, a NULL here would indicate a programming error somewhere else, in which case I'd rather find out about it than gloss over it.

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -810,6 +819,11 @@ class ArnoldAttributes : public IECoreScenePreview::Renderer::AttributesInterfac
 			return m_lightShader.get();
 		}
 
+		const std::vector<ArnoldShaderPtr> lightFilterShaders() const
```
Should this be returning a reference rather than a copy?

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -1416,6 +1431,19 @@ class ArnoldLight : public ArnoldObject
+			// Deal with light filters.
+
+			std::vector<AtNode*> lightFilterNodesVector;
+			std::vector<ArnoldShaderPtr> lightFilterShaders = arnoldAttributes->lightFilterShaders();
```
Again, should this be a const reference?

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -1416,6 +1431,19 @@ class ArnoldLight : public ArnoldObject
+
+			std::vector<AtNode*> lightFilterNodesVector;
+			std::vector<ArnoldShaderPtr> lightFilterShaders = arnoldAttributes->lightFilterShaders();
+			lightFilterNodesVector.reserve( lightFilterShaders.size() );
```
I don't think we actually need this temporary vector, because we could use [`AiArrayAllocate`](https://trac.solidangle.com/arnoldpedia/chrome/site/Arnold-4.0.14.0/doc/api/group__ai__array.html#ga26d7ad1cd1775d85517d017115a74f49) to allocate the array and then use `AiArraySetPtr()` to fill it as we iterate. Best to avoid temporary allocations if possible...

johnhaddon commented on python/GafferSceneUI/ShaderUI.py:
```diff
@@ -142,6 +142,13 @@
+			"nodule:type", "",
+			"plugValueWidget:type", "",
+
+		],
```
We need to add a description here to fix the failing travis tests.

mattigruener commented:
Hey John,

thanks for the feedback - I should have caught most of that myself, sorry.

I've added a couple of new commits that address your feedback and will squash them in once you're happy with them :)

Anything that I missed?

Thanks,

Matti.

mattigruener commented:
I'm not restarting the failed tests because they seem to fail on the Catalogue/Display stuff - not sure if you need more information on those.

johnhaddon commented:
Thanks for the prod Matti - I had intended to look at this yesterday but it escaped me. Thanks for the changes - all looks good. 

One small question - do you think the attribute suffix belongs at the top in the NodeEditor? It feels a little awkward there to me, because I don't think it'll be used all that often and it's not a shader parameter like the others. I wonder what you think to moving it to a separate tab?

Anyway, that's a small one - see what you and the gang think and then maybe sneak in a tweak if you think it makes sense. Either way, could you squash in those final changes and merge?

Thanks!

mattigruener commented:
Hey John,

I've cleaned up the history, but haven't merged yet. I did sneak in a [line](https://github.com/GafferHQ/gaffer/pull/2135/commits/81288eae2b5a7049a001b86c3506a9ddd5d33306) that moves the attributeSuffix over the TabbedWidget. Dan suggested having it there - separated from the shader paramters. How do you feel about that solution?

Matti.

johnhaddon commented:
> I did sneak in a line that moves the attributeSuffix over the TabbedWidget

Yeah, that feels sensible. Thanks Matti!
